src/eos-updater-avahi.c

```diff
@@ -0,0 +1,678 @@
+            void* discoverer_ptr)
+{
+  EosAvahiDiscoverer *discoverer = discoverer_ptr;
+  char address_string[AVAHI_ADDRESS_STR_MAX];
+  GPtrArray *gtxt;
+  EosAvahiService *service;
+  AvahiStringList* iter;
+
+  if (discoverer->state == EOS_AVAHI_FINISHED)
+    return;
+
+  if (!g_hash_table_remove (discoverer->discovered_services, name))
+    {
+      /* maybe it was removed in the meantime */
+      return;
+    }
+
+  gtxt = g_ptr_array_new ();
+  for (iter = txt; iter != NULL; iter = avahi_string_list_get_next (iter))
+    {
+      const gchar* text = (const gchar*)avahi_string_list_get_text (iter);
+      gsize size = avahi_string_list_get_size (iter);
+
+      g_ptr_array_add (gtxt, g_strndup (text, size));
+    }
+  g_ptr_array_add (gtxt, NULL);
+
+  service = g_object_new (EOS_TYPE_AVAHI_SERVICE, NULL);
+  service->name = g_strdup (name);
+  service->domain = g_strdup (domain);
+  service->address = g_strdup (avahi_address_snprint (address_string, sizeof address_string, address));
+  service->port = port;
+  service->txt = (gchar**)g_ptr_array_free (gtxt, FALSE);
+  g_ptr_array_add (discoverer->found_services, service);
+
+  maybe_queue_success_callback (discoverer);
+}
+
+static void
+browse_new (EosAvahiDiscoverer *discoverer,
+            AvahiIfIndex interface,
+            AvahiProtocol protocol,
+            const char *name,
+            const char *type,
+            const char *domain)
+{
+  AvahiServiceResolver *resolver;
+
+  if (discoverer->state == EOS_AVAHI_RESOLVING_ONLY)
+    {
+      return;
+    }
+
+  if (g_hash_table_contains (discoverer->discovered_services, name))
+    {
+      message("name service %s was already found on the network", name);
+      return;
+    }
+
+  resolver = avahi_service_resolver_new (discoverer->client,
+                                         interface,
+                                         protocol,
+                                         name,
+                                         type,
+                                         domain,
+                                         AVAHI_PROTO_UNSPEC,
+                                         0,
+                                         resolve_cb,
+                                         discoverer);
+  if (resolver == NULL)
+    {
+      int failure = avahi_client_errno (discoverer->client);
+
+      queue_error_callback (discoverer,
+                            "Failed to resolve service %s: %s",
+                            name,
+                            avahi_strerror (failure));
+      return;
+    }
+
+  g_hash_table_add (discoverer->discovered_services, g_strdup (name));
+}
+
+static void
+browse_remove (EosAvahiDiscoverer *discoverer,
+               const char *name)
+{
+  guint iter;
+
+  if (g_hash_table_remove (discoverer->discovered_services, name))
+    {
+      maybe_queue_success_callback (discoverer);
+      return;
+    }
+  for (iter = 0; iter < discoverer->found_services->len; ++iter)
```

Comment:
Rather not. We put the name to the `discovered_services` when we learn about its name (in `browse_new` and move it from there to `found_services` when we fetch its data like address, port and so on. So both of them should be disjoint.